python/lsst/ip/diffim/dipoleFitTask.py

```diff
@@ -0,0 +1,921 @@
+        @param xcenPos Desired x-centroid of the positive lobe of the dipole
+        @param ycenPos Desired y-centroid of the positive lobe of the dipole
+        @param xcenNeg Desired x-centroid of the negative lobe of the dipole
+        @param ycenNeg Desired y-centroid of the negative lobe of the dipole
+        @param fluxNeg Desired flux of the negative lobe of the dipole, set to 'flux' if None
+        @param b, x1, y1, xy, x2, y2 Gradient parameters for positive lobe.
+        @see `genBgGradientModel`.
+        @param bNeg, x1Neg, y1Neg, xyNeg, x2Neg, y2Neg Gradient
+        parameters for negative lobe.  @see `genBgGradientModel`. Set
+        to the corresponding positive values if None.
+
+        @param **kwargs Keyword arguments passed through `lmfit` and
+        used by this functor. These must include:
+           @param psf Psf model used to generate the 'star'
+           @param rel_weight Used to signify if positive/negative images are to be
+           included (!= 0. if yes)
+           @param bbox Bounding box containing region to be modelled
+
+        @return numpy.array of width/height matching input bbox,
+        containing dipole model with given centroids and flux(es). If
+        rel_weight = 0., this is a 2-d array with dimensions matching
+        those of bbox; otherwise a stack of three such arrays,
+        representing the dipole (diffim), positive and negative images
+        respectively.
+        """
+
+        psf = kwargs.get('psf')
+        rel_weight = kwargs.get('rel_weight')  # if > 0, we're including pre-sub. images
+        fp = kwargs.get('footprint')
+        bbox = fp.getBBox()
+
+        if fluxNeg is None:
+            fluxNeg = flux
+
+        if self.debug:
+            self.log.log(self.log.DEBUG, '%.2f %.2f %.2f %.2f %.2f %.2f' %
+                         (flux, fluxNeg, xcenPos, ycenPos, xcenNeg, ycenNeg))
+            if x1 is not None:
+                self.log.log(self.log.DEBUG, '     %.2f %.2f %.2f' % (b, x1, y1))
+            if xy is not None:
+                self.log.log(self.log.DEBUG, '     %.2f %.2f %.2f' % (xy, x2, y2))
+
+        posIm = self.genStarModel(bbox, psf, xcenPos, ycenPos, flux)
+        negIm = self.genStarModel(bbox, psf, xcenNeg, ycenNeg, fluxNeg)
+
+        in_x = x
+        if in_x is None:  # use the footprint to generate the input grid
+            y, x = np.mgrid[bbox.getBeginY():bbox.getEndY(), bbox.getBeginX():bbox.getEndX()]
+            in_x = np.array([x, y]) * 1.
+            in_x[0, :] -= in_x[0, :].mean()  # center it!
+            in_x[1, :] -= in_x[1, :].mean()
+
+        gradient = self.genBgGradientModel(in_x, (b, x1, y1, xy, x2, y2))
+        gradientNeg = gradient
+        if bNeg is not None:
+            gradientNeg = self.genBgGradientModel(in_x, (bNeg, x1Neg, y1Neg, xyNeg, x2Neg, y2Neg))
+
+        if gradient is not None:
+            posIm.getArray()[:, :] += gradient
+            negIm.getArray()[:, :] += gradientNeg
```

Comment:
I find the above code confusing. Why do you add `gradient` to `negIm` if `bNeg is None`, else `gradientNeg`? Why do you bother to compute `gradientNeg` at all if `gradient is None`? I think the following would be much cleaner, and it eliminates the need for `genBgGradientModel` to handle the first coefficient being None

        if b is not None:
            gradient = self.genBgGradientModel(in_x, (b, x1, y1, xy, x2, y2))

            # ... code comment explaining why gradientNeg should be gradient if bNeg is None
            if bNeg is not None:
                gradientNeg = self.genBgGradientModel(in_x, (bNeg, x1Neg, y1Neg, xyNeg, x2Neg, y2Neg))
            else:
                gradientNeg = gradient
    
            posIm.getArray()[:, :] += gradient
            negIm.getArray()[:, :] += gradientNeg